### PR TITLE
Updates in train.py rpn layer and predict.py proprocessing

### PR DESCRIPTION
**Pull request by WangLuran**
Changing the nms algorithm in rpn layer by considering inter over area instead of inter over union. It inceases the speed of divergence and mask quality and a little in ap50. The overlapping of  crowns of different types is also solved.
Change in proprocessing in predict.py. Instead of just deleting all crowns with a objectness value less than a threshold, a changed nms algorithm is applied - we use inter over area of masks instead of bounding boxes for a better output.

### CODE REVIEW

**Comment by PatBall1** [comment]
Thanks for this contribution @WangLuran !
There seems to be a few conflicts that are preventing me from merging.
Would you also please comment which functions have been edited (and how)?

**Comment by WangLuran** [comment]
OK, got it! I will merge the commits and write the comments

File: detectree2/models/train.py [hunk 1]
```diff
@@ -391,6 +391,61 @@ def build_train_loader(cls, cfg):
         ),
     )
 
+  # def build_optimizer(cls, cfg, model):
```
Comment (WangLuran) [inline detectree2/models/train.py:394]: using adam optimizer instead of SGD. While SGD is updated recently so this function is commented out.

File: detectree2/models/train.py [hunk 2]
```diff
@@ -448,13 +503,11 @@ def test_train(cls, cfg, model, evaluators=None):
       return results
```
Comment (WangLuran) [inline detectree2/models/train.py:505]: the get_tree_dicts function hsould not be changed, my version is not updated

File: detectree2/models/predict.py [hunk 3]
```diff
@@ -1,64 +1,46 @@
+"""Generate predictions."""
```
Comment (WangLuran) [inline detectree2/models/predict.py:1]: the only change in this one is to import defaultpredictor1 from custom_nms which is a defaultpredictor + proprocessing using mask-nms. The other conflicts are from that my version is not updated and should not be changed

File: detectree2/models/train.py [hunk 4]
```diff
@@ -433,74 +695,62 @@ def load_json_arr(json_path):
...
-    else:
-        cfg.MODEL.WEIGHTS = model_zoo.get_checkpoint_url(base_model)
-
-    cfg.SOLVER.IMS_PER_BATCH = ims_per_batch
-    cfg.SOLVER.BASE_LR = base_lr
-    cfg.SOLVER.MAX_ITER = max_iter
-    cfg.MODEL.ROI_HEADS.NUM_CLASSES = num_classes
-    cfg.TEST.EVAL_PERIOD = eval_period
-    return cfg
-
```
Comment (WangLuran) [inline detectree2/models/train.py:500]: the set_up cfg function is used for the new RPN layer and to load its parameters

File: detectree2/models/train.py [hunk 5]
```diff
@@ -433,74 +695,62 @@ def load_json_arr(json_path):
...
+  cfg.MODEL.PROPOSAL_GENERATOR.NAME  = 'custom_RPN'
+  cfg.nms_thresh_union = 0.2
+  cfg.MODEL.RPN.NMS_THRESH = 0
+
+  #cfg.nms_thresh_union = 0
+  #cfg.MODEL.ROI_HEADS.NMS_THRESH_TEST = 0.1
+
+  #cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.5
+  cfg.MODEL.ROI_HEADS.NUM_CLASSES = 1  # only has one class (pnumonia)
+  return cfg
```
Comment (WangLuran) [inline detectree2/models/train.py:751]: the change in predictions_on_data is simply to change the predictor from DelaultTrainer to the new predictor in custom_nms.py

**Comment by ma595** [comment]
This PR has become stale. @PatBall1 is it still relevant?
